compiler/lake.go

```diff
@@ -49,3 +54,70 @@ func (l *lakeCompiler) NewLakeQuery(pctx *op.Context, program ast.Op, parallelis
 	}
 	return zedruntime.NewQuery(job.pctx, job.Puller(), job.builder.Meters()), nil
 }
+
+func (l *lakeCompiler) NewLakeDeleteQuery(pctx *op.Context, program ast.Op, head *lakeparse.Commitish) (*zedruntime.DeleteQuery, error) {
+	job, err := newDeleteJob(pctx, program, l.src, head)
+	if err != nil {
+		return nil, err
+	}
+	if err := job.Optimize(); err != nil {
+		return nil, err
+	}
+	if err := job.Parallelize(Parallelism); err != nil {
+		return nil, err
+	}
+	if err := job.Build(); err != nil {
+		return nil, err
+	}
+	meters := job.builder.Meters()
+	return zedruntime.NewDeleteQuery(pctx, job.Puller(), meters[0].(zedruntime.DeleteMeter)), nil
+}
+
+type InvalidDeleteWhereQuery struct{}
+
+func (InvalidDeleteWhereQuery) Error() string {
+	return "invalid delete where query: must be a single filter operation"
+}
+
+func newDeleteJob(pctx *op.Context, inAST ast.Op, src *data.Source, head *lakeparse.Commitish) (*Job, error) {
+	parserAST := ast.Copy(inAST)
+	seq, ok := parserAST.(*ast.Sequential)
+	if !ok {
+		return nil, fmt.Errorf("internal error: AST must begin with a Sequential op: %T", parserAST)
+	}
+	if len(seq.Ops) == 0 {
+		return nil, errors.New("internal error: AST Sequential op cannot be empty")
+	}
+	if len(seq.Ops) != 1 {
+		return nil, &InvalidDeleteWhereQuery{}
+	}
+	// add trunk
+	seq.Prepend(&ast.From{
+		Kind: "from",
+		Trunks: []ast.Trunk{{
+			Kind: "Trunk",
+			Source: &ast.Pool{
+				Kind:   "Pool",
+				Delete: true,
+				Spec: ast.PoolSpec{
+					Pool: &ast.String{
+						Kind: "String",
+						Text: "HEAD",
+					},
+				},
+			},
+		}},
+	})
+	entry, err := semantic.Analyze(pctx.Context, seq, src, head)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := entry.Ops[1].(*dag.Filter); !ok {
+		return nil, &InvalidDeleteWhereQuery{}
+	}
```

Comment:
Could we get rid of compiler/kernel.DeleteFilter by doing something like this?
```suggestion
	f, ok := entry.Ops[1].(*dag.Filter)
	if !ok {
		return nil, &InvalidDeleteWhereQuery{}
	}
	// Invert the filter so we keep values that don't match the original.
	f.Expr = &dag.UnaryExpr{
		Kind:    "UnaryExpr",
		Op:      "!",
		Operand: f.Expr,
	}
```

Reply:
We talked about this but this can't be done since adding the not expression here would disable the object filter pass (which doesn't work with not expressions) and would there for require every object to be scanned on a delete that could otherwise be optimized.